ecosystem/indexer-grpc/indexer-grpc-table-info/src/internal_indexer_db_service.rs

```diff
@@ -166,31 +166,30 @@ impl InternalIndexerDBService {
 
     pub async fn run(&mut self, node_config: &NodeConfig) -> Result<()> {
         let mut start_version = self.get_start_version(node_config).await?;
+        let mut target_version = self.db_indexer.main_db_reader.ensure_synced_version()?;
+        let mut step_timer = std::time::Instant::now();
 
         loop {
-            let start_time: std::time::Instant = std::time::Instant::now();
-            let next_version = self.db_indexer.process_a_batch(start_version)?;
-
-            if next_version == start_version {
-                if let Ok(recv_res) =
-                    tokio::time::timeout(Duration::from_millis(100), self.update_receiver.changed())
-                        .await
-                {
-                    if recv_res.is_err() {
-                        info!("update sender is dropped");
-                        return Ok(());
-                    }
+            if target_version == start_version {
+                match self.update_receiver.changed().await {
+                    Ok(_) => {
+                        (step_timer, target_version) = *self.update_receiver.borrow();
+                    },
+                    Err(e) => {
+                        panic!("Failed to get update from update_receiver: {}", e);
+                    },
```

Comment:
This is how this thread knows the main db has quit, right? In that case we should return `Ok` instead of suicide here?

And I realized even in the previous logic, this thread doesn't have a chance to quit until hitting the target version? It can be an issue when the first time the indexer is enabled? (granted we usually quit the maindb by killing the whole process, it'd be better if we deal with this case gracefully, if not too complicated.)

Reply:
1. the DBindexer should have Arc<dyn DbReader> of main_db.  I expect the err should never be caused by main_db quit. (The main db quits after db indexer being dropped). Thus, I think it should panic if this unexpected error occurs.
2.  if the process is killed while the indexer is in the middle of processing a large amount of data. the thread is terminated the same way as other components reading main_db eg: API  is reading transactions using storage interface and the node is killed.

Reply:
ugh, in that case the indexer loop never quits? That should be an issue in tests? 

Shall we implement graceful quitting (separately)? @grao1991 ?

Reply:
I don't disagree.

Reply:
there is method called "run_with_end_version" for testing